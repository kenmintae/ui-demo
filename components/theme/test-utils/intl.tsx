import React, { ReactElement } from 'react';
import { intlShape, IntlProvider } from 'react-intl';
import { mount, render, shallow } from 'enzyme';
import { ThemeProvider as SCThemeProvider } from 'styled-components';

import { ThemeProvider } from '../index';

const defaultMessages = { foobar: 'baz' };
const locale = 'en';

// using example from https://github.com/styled-components/jest-styled-components#theming

/*
Use when styled-components v4 update happens

function getIntlThemeContextParams(options: IntlProvider.Props = {}) {
  const messages = options.messages || defaultMessages;
  const intlProviderInstance = shallow(
    <IntlProvider locale={locale} {...options} messages={messages}>
      <div>empty</div>
    </IntlProvider>,
  ).instance();
  const intlContext = (intlProviderInstance as any).getChildContext();

  return {
    context: {
      ...intlContext,
    },
    childContextTypes: {
      intl: intlShape,
    },
  };
}

export function mountWithThemeIntl(children: ReactElement<any>, options: IntlProvider.Props = {}) {
  return mountWithTheme(children, getIntlThemeContextParams(options));
}

export function renderWithThemeIntl(children: ReactElement<any>, options: IntlProvider.Props = {}) {
  return renderWithTheme(children, getIntlThemeContextParams(options));
}
*/

function getIntlThemeContextParams(options: IntlProvider.Props = {}) {
  const themeProviderInstance = shallow(<ThemeProvider />)
    .dive()
    .instance();
  const themeContext = (themeProviderInstance as any).getChildContext();

  const messages = options.messages || defaultMessages;
  const intlProviderInstance = shallow(
    <IntlProvider locale={locale} {...options} messages={messages}>
      <div>empty</div>
    </IntlProvider>,
  ).instance();
  const intlContext = (intlProviderInstance as any).getChildContext();

  return {
    context: {
      ...themeContext,
      ...intlContext,
    },
    childContextTypes: {
      ...SCThemeProvider.childContextTypes,
      intl: intlShape,
    },
  };
}

export function mountWithThemeIntl(children: ReactElement<any>, options: IntlProvider.Props = {}) {
  return mount(children, getIntlThemeContextParams(options));
}

export function renderWithThemeIntl(children: ReactElement<any>, options: IntlProvider.Props = {}) {
  return render(children, getIntlThemeContextParams(options));
}
