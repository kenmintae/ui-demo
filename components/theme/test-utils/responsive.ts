const minWidthRegex = /min-width: (\d+)px/;
const maxWidthRegex = /max-width: (\d+)px/;

export function stubWindowSize(size: number) {
  window.matchMedia = query => {
    const minWidthMatch = minWidthRegex.exec(query);
    const maxWidthMatch = maxWidthRegex.exec(query);
    const minWidth = minWidthMatch && minWidthMatch[1] ? Number(minWidthMatch[1] as String) : 0;
    const maxWidth = maxWidthMatch && maxWidthMatch[1] ? Number(maxWidthMatch[1] as String) : Infinity;
    return {
      matches: size >= minWidth && size <= maxWidth,
    } as MediaQueryList;
  };
}
