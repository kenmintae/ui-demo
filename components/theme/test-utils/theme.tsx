import React, { ReactElement } from 'react';
import { mount, render as renderEnzyme, shallow } from 'enzyme';
// @ts-ignore
import { ThemeContext, ThemeProvider as SCThemeProvider } from 'styled-components';
import { render, RenderResult } from 'react-testing-library';

import { ThemeProvider } from '../index';

/*
Use this when update to styled-components v4 happens

export function mountWithTheme(children: any, context?: any) {
  if (children.type.prototype instanceof React.Component) {
    const ChildClass = children.type as any;

    class ChildClassWrapper extends ChildClass {
      render() {
        return <ThemeProvider>{super.render()}</ThemeProvider>;
      }
    }

    const ComponentToMount = ChildClassWrapper as any;
    ComponentToMount.defaultProps = ChildClass.defaultProps;
    ComponentToMount.displayName = ChildClass.displayName;

    return mount(<ComponentToMount {...children.props} />, context);
  } else {
    const WrappedStatelessComponent: React.StatelessComponent<any> = (props: any) => (
      <ThemeProvider>{React.cloneElement(children, { ...children.props, ...props })}</ThemeProvider>
    );
    WrappedStatelessComponent.defaultProps = children.type.defaultProps;
    WrappedStatelessComponent.displayName = WrappedStatelessComponent.displayName;

    return mount(<WrappedStatelessComponent {...children.props} />, context);
  }
}
*/

/** @deprecated */
export function mountWithTheme(children: ReactElement<any>) {
  const root = mount(<ThemeProvider>{children}</ThemeProvider>) as any;
  return mount(children, {
    context: root
      .find(SCThemeProvider)
      .instance()
      .getChildContext(),
    childContextTypes: SCThemeProvider.childContextTypes,
  });
}

/** @deprecated */
export function renderWithTheme(children: ReactElement<any>, context?: any) {
  return renderEnzyme(<ThemeProvider>{children}</ThemeProvider>, context);
}

/** @deprecated */
export function shallowWithTheme(children: ReactElement<any>) {
  return shallow(<ThemeProvider>{children}</ThemeProvider>).dive();
}

export function renderWithContext(children: React.ReactNode): RenderResult {
  // TODO: include intl etc too
  return render(<ThemeProvider>{children}</ThemeProvider>);
}
