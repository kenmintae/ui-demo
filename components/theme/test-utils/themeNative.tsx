import React, { ReactElement } from 'react';
import renderer from 'react-test-renderer';

import { ThemeProvider } from '../native';

export function renderWithTheme(children: ReactElement<any>) {
  return renderer.create(<ThemeProvider>{children}</ThemeProvider>);
}
