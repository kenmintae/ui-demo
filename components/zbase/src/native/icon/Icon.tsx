import React, { StatelessComponent } from 'react';
import { Text, TextStyle } from 'react-native';
import { FlattenInterpolation } from 'styled-components';

import { css, withTheme, IconNameString, ThemeInterface } from 'z-frontend-theme/native';
import { fontSizes } from 'z-frontend-theme/utils';

import withUtilPropsNative, { ResultNativeComponentProps } from '../withUtilPropsNative';

export type IconSize = 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge' | 'xxlarge';

export const iconFontSizeMap: { [size in IconSize]: number } = {
  xsmall: 0,
  small: 1,
  medium: 2,
  large: 4,
  xlarge: 5,
  xxlarge: 7,
};

type AdditionalProps = {
  /** Name of the icon to display. Available list: http://zavoloklom.github.io/material-design-iconic-font/icons.html */
  iconName: IconNameString;
  /** Should the icon spin continuously? */
  // spin?: boolean;
  /** Size of the icon. Omit to match the inherited font size. */
  s?: IconSize;
};

export type IconProps = ResultNativeComponentProps<{}, AdditionalProps>;

type InnerProps = { theme: ThemeInterface; style: TextStyle } & AdditionalProps;

// inherit font-size and color
const iconStyle: FlattenInterpolation<InnerProps & AdditionalProps>[] = css`
  font-family: ${props => props.theme.iconFont};
  /* font-style: normal; */ /* uncomment to crash JS VM on start */
  font-size: ${props => (props.s ? fontSizes(iconFontSizeMap[props.s]) : '')};
`;

const IconInner: StatelessComponent<InnerProps> = props => {
  // convert HEX value (like '/f26b') to Int, and then get a char code for it
  // used example https://github.com/oblador/react-native-vector-icons/blob/v4.6.0/lib/generate-icon-set-from-css.js#L16

  const iconStr = String.fromCharCode(parseInt(props.theme.icons[props.iconName].substr(1), 16));
  return <Text style={props.style}>{iconStr}</Text>;
};

export const Icon = withTheme(
  withUtilPropsNative<{}, AdditionalProps>({
    additionalCss: iconStyle,
  })(IconInner),
);

Icon.defaultProps = {
  s: 'large' as 'large',
};
