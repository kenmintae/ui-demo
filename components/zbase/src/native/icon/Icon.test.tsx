import React from 'react';

import { renderWithTheme } from 'z-frontend-theme/test-utils/themeNative';

import { Icon } from './Icon';

describe('Icon component', () => {
  it('should render a Text component with icon as a child', () => {
    const iconRendered = renderWithTheme(<Icon iconName="airplane" s="xlarge" />);
    expect(iconRendered.toJSON()).toMatchSnapshot();
  });
});
