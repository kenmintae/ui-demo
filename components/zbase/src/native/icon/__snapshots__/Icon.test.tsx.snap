// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component should render a Text component with icon as a child 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "fontFamily": "Material-Design-Iconic-Font",
        "fontSize": 24,
      },
      undefined,
    ]
  }
>
  ï„ƒ
</Text>
`;
